irv/irv.py

```diff
@@ -84,22 +83,22 @@ def _populate_ballots_and_candidates(self, text_stream: TextIOBase, permute: boo
             - Whether to randomly permute the order of ballots before processing.
             Potentially useful in testing.  Default False
         """
-        csv_lines_lengths = [len(line) for line in csv.reader(text_stream)]
-        text_stream.seek(0)
-        max_col = max(csv_lines_lengths)
+        self.votes: list[list[str]] = votes
+        for single_ballot in votes:
+            for candidate in single_ballot:
+                if single_ballot.count(candidate) > 1:
+                    votes.remove(single_ballot)
+                    raise ValueError("There are duplicate votes in a single ballot!")
+
+            typecheck = all(isinstance(candidate, str) for candidate in single_ballot)
+            if not typecheck:
+                votes.remove(single_ballot)
+                raise ValueError("Not every value is a string!")
 
-        self.ballots = pd.read_csv(text_stream, header=None, names=range(max_col),
-                                   index_col=False, dtype='str', comment='#', skip_blank_lines=False)
         self.candidates: set = set()
```

Comment:
You have to populate candidates here. I know you opted to not use the RankedChoiceBallots here, but I recommend you recreate the RankedChoiceBallots class, and put a function there that gets all unique candidates, then use it here. As well as the class validations implemented above in the init.